scipy/stats/_binned_statistic.py

```diff
@@ -446,21 +564,31 @@ def binned_statistic_dd(sample, values, statistic='mean',
                 null = np.nan
             np.seterr(**old)
         result.fill(null)
-        for i in np.unique(xy):
-            result[i] = statistic(values[xy == i])
+        for i in np.unique(binnumbers):
+            for vv in range(Vdim):
+                result[vv, i] = statistic(values[vv, binnumbers == i])
 
     # Shape into a proper matrix
-    result = result.reshape(np.sort(nbin))
+    result = result.reshape(np.append(Vdim, np.sort(nbin)))
+
     for i in np.arange(nbin.size):
         j = ni.argsort()[i]
-        result = result.swapaxes(i, j)
+        # Accomodate the extra `Vdim` dimension-zero with `+1`
+        result = result.swapaxes(i+1, j+1)
         ni[i], ni[j] = ni[j], ni[i]
 
     # Remove outliers (indices 0 and -1 for each dimension).
-    core = D * [slice(1, -1)]
+    core = [slice(None)] + Ndim * [slice(1, -1)]
     result = result[core]
 
-    if (result.shape != nbin - 2).any():
+    # Unravel binnumbers into an ndarray, each row the bins for each dimension
+    if(expand_binnumbers and Ndim > 1):
+        binnumbers = np.asarray(np.unravel_index(binnumbers, nbin))
+
+    if np.any(result.shape[1:] != nbin - 2):
         raise RuntimeError('Internal Shape Error')
 
-    return BinnedStatisticddResult(result, edges, xy)
+    # Remove excess dimension-zero if ``Vdim == 1``
+    result = result.squeeze()
```

Comment:
I'm not sure I understand what you mean by 'length-1'.  Because I'm doing ``values = np.atleast_2d(values)``, if the user inputs a `(N,)` array, they would get back a `(1,M)` array (for `N` length input arrays, and `M` bins) --- so I was squeezing that out.  Are you saying that if they input a `(1,N)` array *on purpose*, then they will erroneously get back a `(M,)` array?

To resolve my current understanding of the issue, I've switch to storing the initial number of dimensons, ``initVdim = np.ndim(values)`` in the beginning, and then at the end doing,  

    if(initVdim == 1):
        result = result.reshape(*(nbin-2))

Where `nbin` is actually the number of bins in each dimension `+2` for outlier bins.